Sources/AudioKit/Internals/Utilities/Atomic.swift

```diff
@@ -0,0 +1,81 @@
+//  Atomic.swift
+//  MIDIKit
+//  Created by Steffan Andrews on 2020-12-20.
+
+import Darwin
+
+/// Defines a basic signature to which all locks will conform. Provides the basis for atomic access to stuff.
+private protocol ThreadLock {
+    init()
+
+    /// Lock a resource for writing. So only one thing can write, and nothing else can read or write.
+    func writeLock()
+
+    /// Lock a resource for reading. Other things can also lock for reading at the same time, but nothing else can write at that time.
+    func readLock()
+
+    /// Unlock a resource
+    func unlock()
+}
+
+private final class RWThreadLock: ThreadLock {
+    private var lock = pthread_rwlock_t()
+
+    init() {
+        guard pthread_rwlock_init(&lock, nil) == 0 else {
+            preconditionFailure("Unable to initialize the lock")
+        }
+    }
+
+    deinit {
+        pthread_rwlock_destroy(&lock)
+    }
+
+    func writeLock() {
+        pthread_rwlock_wrlock(&lock)
+    }
+
+    func readLock() {
+        pthread_rwlock_rdlock(&lock)
+    }
+
+    func unlock() {
+        pthread_rwlock_unlock(&lock)
+    }
+}
+
+/// A property wrapper that ensures atomic access to a value, meaning thread-safe with implicit serial read/write access.
+/// Multiple read accesses can potentially read at the same time, just not during a write.
+/// By using `pthread` to do the locking, this safer then using a `DispatchQueue/barrier` as there isn't a chance of priority inversion.
+@propertyWrapper
+public final class Atomic<T> {
```

Comment:
Could we consider using a name that is more specific to the underlying implementation, e.g. `ThreadLockedAccessor`, `LockingThreadAccessor`, `PThreadLockAcccess`, `ThreadLockProperty`, `PThreadBarrier`  ?

These names are not as pretty, but here are few thoughts that I think deserve some consideration:

- It is reasonable to assume the name`Atomic` would represent the full / commonly expected set of features/uses that atomics provide

- It may confuse the very useful mechanism used here with the atomic concepts from C, C++, or even Swift / possible future Swift (I've seen some [swift evolution work](https://swift.org/blog/swift-atomics/), but have not dug into it deeply).

- overloading the name `Atomic` may preclude the use of standardized or 3rd party packages that an AK user may want to include that use `Atomic`

- The property wrapper may/will show up in xcode autocomplete leading to possible unintended use resulting in undefined behavior / headaches for fast-movers or those with less knowledge of multithreading (maybe I missed something in the commits that makes it so this won't happen)

Perhaps you have already thought of all these or other things.
Even so, its pretty cool and instructive to have some github or inline documentation of AK's "framework design thinking"

Reply:
these are good comments, thank you. i didn't actually write that wrapper, but was using it in the Tap class as it seemed preferable to a dispatch queue in this case. it's fine to keep it private to my class as well.

Reply:
Changed to internal `ThreadLockedAccessor` - I would agree that `Atomic` shouldn't be a public name in AK.

Reply:
Gotchya, sounds good!

Reply:
thanks!